ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY with relative local dependency (difference between windows and linux CI?)

### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

9.0.2

### Which area(s) of pnpm are affected? (leave empty if unsure)

Lockfile

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/Ryex/ic10emu/actions/runs/8731724803/job/23957592473

### Reproduction steps

- add a dependency in the form of `"ic10emu_wasm": "file:../ic10emu_wasm/pkg",`
-  create lockfile using fresh install of 9.0.2 and empty `node_modules` with `pnpm install` under windows, no issues
- push to github
- use workflow to install latest pnpm, get `v9.0.2`
- trigger `pnpm install` in workflow
- get

```
Lockfile is up to date, resolution step is skipped
 ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY  Broken lockfile: no entry for 'ic10emu_wasm@file:../ic10emu_wasm/pkg' in pnpm-lock.yaml
```

 - look at [lockfile](https://github.com/Ryex/ic10emu/blob/main/www/pnpm-lock.yaml) and see no issues

```yaml
 dependencies:
  // ...
  ic10emu_wasm:
    specifier: file:../ic10emu_wasm/pkg
    version: file:../ic10emu_wasm/pkg
  ic10lsp_wasm:
    specifier: file:../ic10lsp_wasm/pkg
    version: file:../ic10lsp_wasm/pkg
   // ...
   
packages:

   // ...

   file:../ic10emu_wasm/pkg:
    resolution: {directory: ../ic10emu_wasm/pkg, type: directory}
    name: ic10emu_wasm
    dev: false

  file:../ic10lsp_wasm/pkg:
    resolution: {directory: ../ic10lsp_wasm/pkg, type: directory}
    name: ic10lsp_wasm
    dev: false
    
```

### Describe the Bug

pnpm says lockfile missing a entry for a dependency specified as a relative directory
inspection of that lockfile reveals that the entry for the dependency is present

### Expected Behavior

`pnpm install` to work with no issues as it has previously

### Which Node.js version are you using?

windows:  v21.7.1  action runner: v18.20.2

### Which operating systems have you used?

- [ ] macOS
- [X] Windows
- [X] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

Ubuntu   22.04.4   LTS (guthub action runner)

--- a/.changeset/cold-pears-relax.md
+++ b/.changeset/cold-pears-relax.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/lockfile-file": patch
+"pnpm": patch
+---
+
+Lockfiles with local directory dependencies should be successfully converted to the new format [#7955](https://github.com/pnpm/pnpm/issues/7955).

--- a/.changeset/cold-pears-relax.md
+++ b/.changeset/cold-pears-relax.md
@@ -3,4 +3,4 @@
 "pnpm": patch
 ---
 
-Lockfiles with local directory dependencies should be successfully converted to the new format [#7955](https://github.com/pnpm/pnpm/issues/7955).
+Lockfiles with local or git-hosted dependencies are now successfully converted to the new lockfile format [#7955](https://github.com/pnpm/pnpm/issues/7955).

--- a/lockfile/lockfile-file/src/lockfileFormatConverters.ts
+++ b/lockfile/lockfile-file/src/lockfileFormatConverters.ts
@@ -202,9 +202,9 @@ function convertPkgIds (lockfile: LockfileFile): void {
   for (const [pkgId, pkg] of Object.entries(lockfile.packages ?? {})) {
     if (pkg.name) {
       const { id, peersSuffix } = parseDepPath(pkgId)
-      let newId: string
+      let newId = `${pkg.name}@`
       if ('tarball' in pkg.resolution) {
-        newId = pkg.resolution.tarball
+        newId += pkg.resolution.tarball
         if (pkg.resolution.path) {
           newId += `#path:${pkg.resolution.path}`
         }
@@ -214,7 +214,7 @@ function convertPkgIds (lockfile: LockfileFile): void {
           newId += `&path:${pkg.resolution.path}`
         }
       } else if ('directory' in pkg.resolution) {
-        newId = `${pkg.name}@${id}`
+        newId += id
       } else {
         continue
       }

--- a/lockfile/lockfile-file/src/lockfileFormatConverters.ts
+++ b/lockfile/lockfile-file/src/lockfileFormatConverters.ts
@@ -209,7 +209,7 @@ function convertPkgIds (lockfile: LockfileFile): void {
           newId += `#path:${pkg.resolution.path}`
         }
       } else if ('repo' in pkg.resolution) {
-        newId = `${pkg.resolution.repo.startsWith('git+') ? '' : 'git+'}${pkg.resolution.repo}#${pkg.resolution.commit}`
+        newId += `${pkg.resolution.repo.startsWith('git+') ? '' : 'git+'}${pkg.resolution.repo}#${pkg.resolution.commit}`
         if (pkg.resolution.path) {
           newId += `&path:${pkg.resolution.path}`
         }

--- a/lockfile/lockfile-file/src/lockfileFormatConverters.ts
+++ b/lockfile/lockfile-file/src/lockfileFormatConverters.ts
@@ -201,6 +201,7 @@ function convertPkgIds (lockfile: LockfileFile): void {
   if (lockfile.packages == null || isEmpty(lockfile.packages)) return
   for (const [pkgId, pkg] of Object.entries(lockfile.packages ?? {})) {
     if (pkg.name) {
+      const { id, peersSuffix } = parseDepPath(pkgId)
       let newId: string
       if ('tarball' in pkg.resolution) {
         newId = pkg.resolution.tarball
@@ -212,10 +213,11 @@ function convertPkgIds (lockfile: LockfileFile): void {
         if (pkg.resolution.path) {
           newId += `&path:${pkg.resolution.path}`
         }
+      } else if ('directory' in pkg.resolution) {
+        newId = `${pkg.name}@${id}`
       } else {
         continue
       }
-      const { id, peersSuffix } = parseDepPath(pkgId)
       oldIdToNewId[pkgId] = `${newId}${peersSuffix}`
       if (id !== pkgId) {
         oldIdToNewId[id] = newId
@@ -242,7 +244,11 @@ function convertPkgIds (lockfile: LockfileFile): void {
     for (const depType of ['dependencies', 'optionalDependencies'] as const) {
       for (const [alias, depPath] of Object.entries(pkg[depType] ?? {})) {
         if (oldIdToNewId[depPath]) {
-          pkg[depType]![alias] = oldIdToNewId[depPath]
+          if (oldIdToNewId[depPath].startsWith(`${alias}@`)) {
+            pkg[depType]![alias] = oldIdToNewId[depPath].substring(alias.length + 1)
+          } else {
+            pkg[depType]![alias] = oldIdToNewId[depPath]
+          }
         }
       }
     }
@@ -252,7 +258,11 @@ function convertPkgIds (lockfile: LockfileFile): void {
     for (const depType of ['dependencies', 'optionalDependencies', 'devDependencies'] as const) {
       for (const [alias, { version }] of Object.entries(importer[depType] ?? {})) {
         if (oldIdToNewId[version]) {
-          importer[depType]![alias].version = oldIdToNewId[version]
+          if (oldIdToNewId[version].startsWith(`${alias}@`)) {
+            importer[depType]![alias].version = oldIdToNewId[version].substring(alias.length + 1)
+          } else {
+            importer[depType]![alias].version = oldIdToNewId[version]
+          }
         }
       }
     }

--- a/lockfile/lockfile-file/test/lockfileFormatConverters.test.ts
+++ b/lockfile/lockfile-file/test/lockfileFormatConverters.test.ts
@@ -40,3 +40,48 @@ test('convertToLockfileObject converts directory dependencies', () => {
     },
   })
 })
+
+test('convertToLockfileObject converts git-hosted dependencies', () => {
+  expect(convertToLockfileObject({
+    lockfileVersion: '',
+    importers: {
+      '.': {
+        dependencies: {
+          'is-negative': {
+            specifier: 'github:kevva/is-negative',
+            version: 'github.com/kevva/is-negative/1d7e288222b53a0cab90a331f1865220ec29560c',
+          },
+        },
+      },
+    },
+    packages: {
+      'github.com/kevva/is-negative/1d7e288222b53a0cab90a331f1865220ec29560c': {
+        resolution: { tarball: 'https://codeload.github.com/kevva/is-negative/tar.gz/1d7e288222b53a0cab90a331f1865220ec29560c' },
+        name: 'is-negative',
+        version: '2.1.0',
+        engines: { node: '>=0.10.0' },
+        dev: false,
+      } as any, // eslint-disable-line
+    },
+  })).toMatchObject({
+    lockfileVersion: '',
+    importers: {
+      '.': {
+        dependencies: {
+          'is-negative': 'https://codeload.github.com/kevva/is-negative/tar.gz/1d7e288222b53a0cab90a331f1865220ec29560c',
+        },
+        specifiers: {
+          'is-negative': 'github:kevva/is-negative',
+        },
+      },
+    },
+    packages: {
+      'is-negative@https://codeload.github.com/kevva/is-negative/tar.gz/1d7e288222b53a0cab90a331f1865220ec29560c': {
+        resolution: { tarball: 'https://codeload.github.com/kevva/is-negative/tar.gz/1d7e288222b53a0cab90a331f1865220ec29560c' },
+        name: 'is-negative',
+        version: '2.1.0',
+        engines: { node: '>=0.10.0' },
+      },
+    },
+  })
+})

--- a/lockfile/lockfile-file/test/lockfileFormatConverters.test.ts
+++ b/lockfile/lockfile-file/test/lockfileFormatConverters.test.ts
@@ -85,3 +85,50 @@ test('convertToLockfileObject converts git-hosted dependencies', () => {
     },
   })
 })
+
+test('convertToLockfileObject converts git-hosted dependencies via ssh', () => {
+  expect(convertToLockfileObject({
+    lockfileVersion: '',
+    importers: {
+      '.': {
+        dependencies: {
+          'git-resolver': {
+            specifier: 'ssh://git@gitlab:pnpm/git-resolver#988c61e11dc8d9ca0b5580cb15291951812549dc',
+            version: 'git+ssh://git@gitlab/pnpm/git-resolver#988c61e11dc8d9ca0b5580cb15291951812549dc',
+          },
+        },
+      },
+    },
+    packages: {
+      'git+ssh://git@gitlab/pnpm/git-resolver#988c61e11dc8d9ca0b5580cb15291951812549dc': {
+        name: 'git-resolver',
+        resolution: {
+          commit: '988c61e11dc8d9ca0b5580cb15291951812549dc',
+          repo: 'ssh://git@gitlab/pnpm/git-resolver',
+          type: 'git',
+        },
+      } as any, // eslint-disable-line
+    },
+  })).toMatchObject({
+    lockfileVersion: '',
+    importers: {
+      '.': {
+        dependencies: {
+          'git-resolver': 'git+ssh://git@gitlab/pnpm/git-resolver#988c61e11dc8d9ca0b5580cb15291951812549dc',
+        },
+        specifiers: {
+          'git-resolver': 'ssh://git@gitlab:pnpm/git-resolver#988c61e11dc8d9ca0b5580cb15291951812549dc',
+        },
+      },
+    },
+    packages: {
+      'git-resolver@git+ssh://git@gitlab/pnpm/git-resolver#988c61e11dc8d9ca0b5580cb15291951812549dc': {
+        resolution: {
+          commit: '988c61e11dc8d9ca0b5580cb15291951812549dc',
+          repo: 'ssh://git@gitlab/pnpm/git-resolver',
+          type: 'git',
+        },
+      },
+    },
+  })
+})

--- a/lockfile/lockfile-file/test/lockfileFormatConverters.test.ts
+++ b/lockfile/lockfile-file/test/lockfileFormatConverters.test.ts
@@ -1,0 +1,42 @@
+import { convertToLockfileObject } from '../lib/lockfileFormatConverters'
+
+test('convertToLockfileObject converts directory dependencies', () => {
+  expect(convertToLockfileObject({
+    lockfileVersion: '',
+    importers: {
+      '.': {
+        dependencies: {
+          a: {
+            specifier: 'file:../a',
+            version: 'file:../a',
+          },
+        },
+      },
+    },
+    packages: {
+      'file:../a': {
+        resolution: { directory: '../a', type: 'directory' },
+        name: 'a',
+        dev: false,
+      } as any, // eslint-disable-line
+    },
+  })).toMatchObject({
+    lockfileVersion: '',
+    importers: {
+      '.': {
+        dependencies: {
+          a: 'file:../a',
+        },
+        specifiers: {
+          a: 'file:../a',
+        },
+      },
+    },
+    packages: {
+      'a@file:../a': {
+        resolution: { directory: '../a', type: 'directory' },
+        name: 'a',
+      },
+    },
+  })
+})

--- a/lockfile/lockfile-file/test/lockfileV6Converters.test.ts
+++ b/lockfile/lockfile-file/test/lockfileV6Converters.test.ts
@@ -236,7 +236,7 @@ test('convertToLockfileObject() converts package IDs', () => {
       project1: {
         dependencies: {
           'is-positive': 'https://codeload.github.com/kevva/is-positive/tar.gz/97edff6f525f192a3f83cea1944765f769ae2678(@babel/core@2.0.0)',
-          tarball: 'https://registry.npmjs.org/is-positive/-/is-positive-1.0.0.tgz',
+          tarball: 'is-positive@https://registry.npmjs.org/is-positive/-/is-positive-1.0.0.tgz',
           'git-hosted': 'git+ssh://git@gitlab/pnpm/git-resolver#988c61e11dc8d9ca0b5580cb15291951812549dc(foo@1.0.0)',
           'is-odd': '1.0.0',
         },
@@ -249,16 +249,16 @@ test('convertToLockfileObject() converts package IDs', () => {
       },
     },
     packages: {
-      'https://codeload.github.com/kevva/is-positive/tar.gz/97edff6f525f192a3f83cea1944765f769ae2678(@babel/core@2.0.0)': {
+      'is-positive@https://codeload.github.com/kevva/is-positive/tar.gz/97edff6f525f192a3f83cea1944765f769ae2678(@babel/core@2.0.0)': {
         name: 'is-positive',
         resolution: { tarball: 'https://codeload.github.com/kevva/is-positive/tar.gz/97edff6f525f192a3f83cea1944765f769ae2678' },
       },
-      'https://registry.npmjs.org/is-positive/-/is-positive-1.0.0.tgz': {
+      'is-positive@https://registry.npmjs.org/is-positive/-/is-positive-1.0.0.tgz': {
         name: 'is-positive',
         resolution: { tarball: 'https://registry.npmjs.org/is-positive/-/is-positive-1.0.0.tgz', integrity: '' },
       },
-      'git+ssh://git@gitlab/pnpm/git-resolver#988c61e11dc8d9ca0b5580cb15291951812549dc(foo@1.0.0)': {
-        id: 'git+ssh://git@gitlab/pnpm/git-resolver#988c61e11dc8d9ca0b5580cb15291951812549dc',
+      'git-hosted@git+ssh://git@gitlab/pnpm/git-resolver#988c61e11dc8d9ca0b5580cb15291951812549dc(foo@1.0.0)': {
+        id: 'git-hosted@git+ssh://git@gitlab/pnpm/git-resolver#988c61e11dc8d9ca0b5580cb15291951812549dc',
         name: 'git-hosted',
         resolution: {
           commit: '988c61e11dc8d9ca0b5580cb15291951812549dc',

